### PR TITLE
Google_oauth2 sign in added

### PR DESCRIPTION
**Pull request by Charchit507**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -57,4 +57,8 @@ gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
 
 gem 'devise'
 
+gem 'omniauth-rails_csrf_protection'
```
Comment (neeteshgupta95) [inline Gemfile]: why do we need this?

File: app/models/user.rb [hunk 2]
```diff
@@ -5,5 +5,22 @@ def full_name
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :validatable, :trackable
+         :recoverable, :rememberable, :validatable, :trackable, :omniauthable, omniauth_providers: [:google_oauth2]
+
+  def self.from_omniauth(access_token)
+    user = User.where(email: access_token.info.email).first
```
Comment (neeteshgupta95) [inline app/models/user.rb]: use `find_by` instead of where and first

File: db/migrate/20220831123022_add_omniauth_to_users.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class AddOmniauthToUsers < ActiveRecord::Migration[6.0]
+  def change
+    add_column :users, :name, :string
```
Comment (neeteshgupta95) [inline db/migrate/20220831123022_add_omniauth_to_users.rb]: why this column needed?

File: db/migrate/20220831123022_add_omniauth_to_users.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+    add_column :users, :image, :string
```
Comment (neeteshgupta95) [inline db/migrate/20220831123022_add_omniauth_to_users.rb]: why this column needed? Are you storing image?

File: app/models/user.rb [hunk 5]
```diff
@@ -5,5 +5,22 @@ def full_name
...
+    unless user
+      user = User.create(
+        email: access_token.info.email,
+        password: Devise.friendly_token[0,20]
+      )
+    end
+    user.name = access_token.info.name
```
Comment (neeteshgupta95) [inline app/models/user.rb]: store value in first_name and last_name

File: app/models/user.rb [hunk 6]
```diff
@@ -5,5 +5,22 @@ def full_name
...
+    user.image = access_token.info.image
```
Comment (neeteshgupta95) [inline app/models/user.rb]: remove that as we are not storing image as of now

**Review by neeteshgupta95 — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 1 above)
Comment (Charchit507) [reply Gemfile]: sir ye authenticity error nhi aaye uske liye hai

File: db/seeds.rb [hunk 7]
```diff
@@ -5,3 +5,12 @@
...
 #   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
 #   Character.create(name: 'Luke', movie: movies.first)
+
+
+100.times do |i|
+  puts "creating student #{i+1}"
+  Article.create(
+    title: "This is #{i+1} Article",
+    text: "Hey I am #{i+1} text"
+  )
```
Comment (neeteshgupta95) [inline db/seeds.rb]: do not hit the database again and again, create article in bulk

**Review by neeteshgupta95 — CHANGES_REQUESTED** [review]

File: db/seeds.rb [hunk 8]
```diff
@@ -7,10 +7,18 @@
 #   Character.create(name: 'Luke', movie: movies.first)
 
 
-100.times do |i|
-  puts "creating student #{i+1}"
-  Article.create(
-    title: "This is #{i+1} Article",
-    text: "Hey I am #{i+1} text"
-  )
+[].tap do |array|
```
Comment (neeteshgupta95) [inline db/seeds.rb]: why do we need this block?

File: Gemfile [hunk 9]
```diff
@@ -57,7 +57,7 @@ gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
 
 gem 'devise'
 
-# gem 'omniauth-rails_csrf_protection'
+gem 'omniauth-rails_csrf_protection'
```
Comment (neeteshgupta95) [inline Gemfile]: Do google authentication without this gem

**Review by neeteshgupta95 — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 9 above)
Comment (Charchit507) [reply Gemfile]: sir iss gem ke bina ye work nhi kr rha

Comment (Charchit507) [reply Gemfile]: it shows me authenticity error

File: db/seeds.rb [hunk 10]
```diff
@@ -5,3 +5,36 @@
...
+
+100.times do |i|
+  created_at = Time.current
+  result = Article.insert_all(
+    [
+      {
+        title: "This is #{i+1} Article",
+        text: "This is #{i+1} Text",
+        created_at: created_at,
+        updated_at: created_at
```
Comment (neeteshgupta95) [inline db/seeds.rb]: why these line?

File: Gemfile (hunk 1 above)
Comment (neeteshgupta95) [reply Gemfile]: use these gem 

```
gem 'omniauth', '~> 1.9.1'
gem 'omniauth-google-oauth2'
```

File: db/seeds.rb [hunk 11]
```diff
@@ -5,3 +5,36 @@
...
+#       updated_at: created_at
+#     }
+#   end
+
+#   result = Article.insert_all array
+#   puts result.inspect
+
+100.times do |i|
+  created_at = Time.current
+  result = Article.insert_all(
```
Comment (neeteshgupta95) [inline db/seeds.rb]: its not creating in bulk

File: db/seeds.rb [hunk 12]
```diff
@@ -5,3 +5,36 @@
...
+    [
+      {
+        title: "This is #{i+1} Article",
+        text: "This is #{i+1} Text",
```
Comment (neeteshgupta95) [inline db/seeds.rb]: create meaningful name of article and description

Comment (neeteshgupta95) [reply db/seeds.rb]: Use this fake data https://deeporion.slack.com/archives/C03QRV1H1GR/p1662395344111339

**Review by neeteshgupta95 — CHANGES_REQUESTED** [review]

File: db/seeds.rb (hunk 10 above)
Comment (Charchit507) [reply db/seeds.rb]: these lines are used to create timestamps

Comment (neeteshgupta95) [reply db/seeds.rb]: Why do you writing it own?

File: Gemfile [hunk 13]
```diff
@@ -8,7 +8,7 @@ gem 'pg', '~> 1.4', '>= 1.4.2'
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails', branch: 'main'
 gem 'rails', '~> 6.0.5', '>= 6.0.5.1'
 # Use sqlite3 as the database for Active Record
-gem 'sqlite3', '~> 1.4'
+# gem 'sqlite3', '~> 1.4'
```
Comment (neeteshgupta95) [inline Gemfile]: remove that completly

File: Gemfile [hunk 14]
```diff
@@ -57,4 +57,11 @@ gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
 
 gem 'devise'
 
+# gem 'omniauth-rails_csrf_protection'
```
Comment (neeteshgupta95) [inline Gemfile]: remove that completly

File: Gemfile [hunk 15]
```diff
@@ -57,4 +57,11 @@ gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
...
+
 gem 'kaminari'
+
+# gem 'omniauth-google-oauth2'
```
Comment (neeteshgupta95) [inline Gemfile]: remove that completly, its duplicate

File: app/controllers/application_controller.rb [hunk 16]
```diff
@@ -6,5 +6,6 @@ class ApplicationController < ActionController::Base
 
   def configure_permitted_parameters
     devise_parameter_sanitizer.permit(:sign_up, keys: [:first_name, :last_name])
+    devise_parameter_sanitizer.permit(:account_update, keys: [:first_name, :last_name])
```
Comment (neeteshgupta95) [inline app/controllers/application_controller.rb:9]: why is this needed to account_update?

File: app/views/layouts/application.html.erb [hunk 17]
```diff
@@ -6,6 +6,7 @@
     <%= csp_meta_tag %>
 
     <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/js/bootstrap.bundle.min.js" integrity="sha384-MrcW6ZMFYlzcLA8Nl+NtUVF0sA7MsXsP1UyJoMp4YLEuNSfAP+JcXn/tWtIaxVXM" crossorigin="anonymous"></script>
```
Comment (neeteshgupta95) [inline app/views/layouts/application.html.erb:9]: why we are using it with cdn?

**Review by neeteshgupta95 — CHANGES_REQUESTED** [review]

File: db/seeds.rb [hunk 18]
```diff
@@ -5,3 +5,1045 @@
...
+#         created_at: created_at,
+#         updated_at: created_at
+#       }
+#     ]
+#   )
+# end
+
+# puts result.inspect
+
+# puts Article.count
```
Comment (neeteshgupta95) [inline db/seeds.rb]: remove that

**Review by neeteshgupta95 — CHANGES_REQUESTED** [review]

File: db/migrate/20220803091536_create_articles.rb [hunk 19]
```diff
@@ -4,7 +4,7 @@ def change
       t.string :title
       t.text :text
 
-      t.timestamps
+      t.timestamps default: -> { 'CURRENT_TIMESTAMP' }
```
Comment (neeteshgupta95) [inline db/migrate/20220803091536_create_articles.rb]: remove that CURRENT_TIMESTAMP

File: db/schema.rb [hunk 20]
```diff
@@ -18,8 +18,8 @@
   create_table "articles", force: :cascade do |t|
     t.string "title"
     t.text "text"
-    t.datetime "created_at", precision: 6, null: false
-    t.datetime "updated_at", precision: 6, null: false
+    t.datetime "created_at", precision: 6, default: -> { "CURRENT_TIMESTAMP" }, null: false
+    t.datetime "updated_at", precision: 6, default: -> { "CURRENT_TIMESTAMP" }, null: false
   end
```
Comment (neeteshgupta95) [inline db/schema.rb]: remove CURRENT_TIMESTAMP from here as well

File: db/seeds.rb [hunk 21]
```diff
@@ -5,3 +5,1045 @@
...
+
+result = Article.insert_all(
+  [
+    {"title":"Geological Engineer","text":"orci pede venenatis non sodales sed tincidunt eu felis fusce posuere felis sed lacus morbi sem mauris"},
+    {"title":"Registered Nurse","text":"vitae mattis nibh ligula nec sem duis aliquam convallis nunc proin at turpis a pede posuere"},
+    {"title":"Statistician III","text":"primis in faucibus orci luctus et ultrices posuere cubilia curae duis faucibus accumsan odio curabitur convallis duis consequat"},
+    {"title":"Web Developer III","text":"tempor convallis nulla neque libero convallis eget eleifend luctus ultricies eu nibh quisque id justo sit amet sapien"},
+    {"title":"Programmer Analyst IV","text":"sapien sapien non mi integer ac neque duis bibendum morbi"},
```
Comment (neeteshgupta95) [inline db/seeds.rb]: @charchitk01 Its a bug for having not null issue. We will create articles one by one
Make changes like this 
```suggestion
Article.create(
  [
    {title: "Geological Engineer", text: "orci pede venenatis non sodales sed tincidunt eu felis fusce posuere felis sed lacus morbi sem mauris"},
    {title: "Registered Nurse", text: "vitae mattis nibh ligula nec sem duis aliquam convallis nunc proin at turpis a pede posuere" }
  ]
)
```

**Review by neeteshgupta95 — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 16 above)
Comment (Charchit507) [reply app/controllers/application_controller.rb:9]: when i try to update my account is shows me first name and last name unpermitted

File: app/views/layouts/application.html.erb (hunk 17 above)
Comment (Charchit507) [reply app/views/layouts/application.html.erb:9]: cdn is used for activate dropdown menu of bootstrap ( js functionality )

File: db/seeds.rb (hunk 21 above)
Comment (Charchit507) [reply db/seeds.rb]: okay sir

File: app/views/layouts/application.html.erb (hunk 17 above)
Comment (neeteshgupta95) [reply app/views/layouts/application.html.erb:9]: @Charchit507 use alternate way instead of cdn

**Review by neeteshgupta95 — APPROVED** [review]
